hail/src/main/scala/is/hail/expr/types/physical/PType.scala

```diff
@@ -114,6 +110,51 @@ object PType {
   val genInsertable: Gen[PStruct] = genInsertableStruct
 
   implicit def arbType = Arbitrary(genArb)
+
+  def canonical(t: Type): PType = {
+    t match {
+      case t: TInt32 => PInt32(t.required)
+      case t: TInt64 => PInt64(t.required)
+      case t: TFloat32 => PFloat32(t.required)
+      case t: TFloat64 => PFloat64(t.required)
+      case t: TBoolean => PBoolean(t.required)
+      case t: TBinary => PBinary(t.required)
+      case t: TString => PString(t.required)
+      case t: TCall => PCall(t.required)
+      case t: TLocus => PLocus(t.rg, t.required)
+      case t: TInterval => PInterval(canonical(t.pointType), t.required)
+      case t: TArray => PArray(canonical(t.elementType), t.required)
+      case t: TSet => PSet(canonical(t.elementType), t.required)
+      case t: TDict => PDict(canonical(t.keyType), canonical(t.valueType), t.required)
+      case t: TTuple => PTuple(t.types.map(canonical), t.required)
+      case t: TStruct => PStruct(t.fields.map(f => PField(f.name, canonical(f.typ), f.index)), t.required)
+      case t: TNDArray => PNDArray(canonical(t.elementType), t.required)
+    }
+  }
+
+  // currently identity
+  def canonical(t: PType): PType = {
```

Comment:
How is this not just `canonical(t.virtualType)`?

Reply:
requiredness. Canonical PTypes will still have a notion of requiredness, and I want this to work when we strip requiredness out of virtual types.

Reply:
but currently they are the same

Reply:
Ah, got it, thanks for clarifying!